qiskit/transpiler/passes/routing/algorithms/bip_model.py

```diff
@@ -0,0 +1,418 @@
+        for t in range(self.depth - 1):
+            # This is a dummy time step and the next one is dummy too
+            if self._is_dummy_step(t) and self._is_dummy_step(t + 1):
+                # We cannot use the next time step unless this one is used too
+                mdl.add_constraint(z[t] >= z[t + 1], ctname=f"dummy_precedence_{t}")
+        # Symmetry breaking on the line -- only works on line topology!
+        if line_symm:
+            for h in range(1, self.num_vqubits):
+                mdl.add_constraint(
+                    sum(w[0, p, 0] for p in range(h))
+                    + sum([w[0, q, self.num_pqubits - 1] for q in range(h, self.num_vqubits)])
+                    >= 1,
+                    ctname=f"sym_break_line_{h}",
+                )
+
+        # *** Define objevtive function ***
+        if objective == "depth":
+            objexr = sum(z[t] for t in range(self.depth) if self._is_dummy_step(t))
+            for t in range(self.depth - 1):
+                for q in range(self.num_vqubits):
+                    for (i, j) in self._arcs:
+                        objexr += 0.1 * x[t, q, i, j]
+            mdl.minimize(objexr)
+        elif objective == "error_rate":
+            self._set_error_rate_obj(mdl)
+        elif objective == "balanced":
+            self._set_balanced_obj(mdl)
+        else:
+            raise TranspilerError(f"Unknown objective type: {objective}")
+
+        self.problem = mdl
+        logger.info("BIP problem stats: %s", self.problem.statistics)
+
+    @staticmethod
+    def _set_error_rate_obj(model):
+        """Set the minimum error rate objective function."""
+        raise NotImplementedError("objective: 'error_rate' is not implemented")
+
+    @staticmethod
+    def _set_balanced_obj(model):
+        """Set the minimum balanced (weighted sum of error_rate and depth) objective function."""
+        raise NotImplementedError("objective: 'balanced' is not implemented")
+
+    def solve_cpx_problem(self, time_limit: float = 60, heuristic_emphasis: bool = False):
+        """Solve the Cplex model.
+
+        Args:
+            time_limit:
+                Time limit (seconds) given to Cplex.
+
+            heuristic_emphasis:
+                Focus on getting good solutions rather than proving optimality
+
+        Raises:
+            MissingOptionalLibraryError: If cplex is not installed
+            TranspilerError: if fails to solve the Cplex problem within given timelimit.
+        """
+        if not HAS_CPLEX:
+            raise MissingOptionalLibraryError(
+                libname="CPLEX",
+                name="CplexOptimizer",
+                pip_install="pip install cplex",
+            )
+        self.problem.set_time_limit(time_limit)
+        self.problem.context.cplex_parameters.randomseed = 777
+        # self.problem.context.cplex_parameters.threads = 4
+        if heuristic_emphasis:
+            self.problem.context.cplex_parameters.emphasis.mip = 5
```

Comment:
If you want users to control empasis, I recommend to use the native value {-1,...,5} with the [reference](https://www.ibm.com/docs/en/icos/20.1.0?topic=parameters-mip-emphasis-switch).

Reply:
mip emphasis 5 is newly introduced for 20.1. So, it raises an error for cplex 12.10 users. So, even if you make `heuristic_emphasis`, I recommend to use values 1 or 4.
https://www.ibm.com/docs/en/icos/12.10.0?topic=parameters-mip-emphasis-switch

Reply:
Passing a dict as `cplex_parameters` to `solve` is another option for users to control cplex.
https://ibmdecisionoptimization.github.io/docplex-doc/mp/docplex.mp.model.html#docplex.mp.model.Model.solve

Reply:
7f4c6a1